DDInstall/DDInstall.psm1

```diff
@@ -0,0 +1,215 @@
+
+    .LINK
+    Learn more about the Datadog Windows Agent User:
+    https://docs.datadoghq.com/agent/guide/windows-agent-ddagent-user/
+    #>
+}
+
+###############################
+# Unexported helper functions #
+###############################
+
+function formatAgentInstallerParameters($params)
+{
+    $formattedparams = "APIKEY=$($params.ApiKey)"
+
+    if ($params.ContainsKey('Site'))                        { $formattedparams += " SITE=$($params.Site)"}
+    if ($params.ContainsKey('Tags'))                        { $formattedparams += " TAGS=``$($params.Tags)``" }
+    if ($params.ContainsKey('Hostname'))                    { $formattedparams += " HOSTNAME=$($params.Hostname)" }
+    if ($params.ContainsKey('DDAgentUsername'))             { $formattedparams += " DDAGENTUSER_NAME=$($params.DDAgentUsername)" }
+    if ($params.ContainsKey('DDAgentPassword'))             { $formattedparams += " DDAGENTUSER_PASSWORD=$($params.DDAgentPassword)" }
+    if ($params.ContainsKey('ApplicationDataDirectory'))    { $formattedparams += " APPLICATIONDATADIRECTORY=$($params.ApplicationDataDirectory)" }
+    if ($params.ContainsKey('ProjectLocation'))             { $formattedparams += " PROJECTLOCATION=$($params.ProjectLocation)" }
+
+    return $formattedparams
+}
+
+function downloadAsset($url, $outFile)
+{
+    # Suppress progress bar to prevent performance issues (affects older windows versions, see https://github.com/PowerShell/PowerShell/issues/13414)
+    $ProgressPreference = 'SilentlyContinue'
+
+    Invoke-WebRequest -UseBasicParsing -Uri $url -OutFile $outFile
+    if (-not $?)
+    {
+        $exception = [Exception]::new("Download failed.")
+        $PSCmdlet.ThrowTerminatingError([System.Management.Automation.ErrorRecord]::new($exception, "FatalError", [Management.Automation.ErrorCategory]::InvalidOperation, $null))
+    }
+
+    if (-not (hasValidDDSignature -asset $outFile))
+    {
+        $exception = [Exception]::new("$outFile did not pass the signature check.")
+        $PSCmdlet.ThrowTerminatingError([System.Management.Automation.ErrorRecord]::new($exception, "FatalError", [Management.Automation.ErrorCategory]::InvalidOperation, $null))
+    }
+}
+
+function hasValidDDSignature($asset)
+{
+    if (-not (Test-Path $asset -PathType Leaf)) {
+        $exception = [Exception]::new("$asset not found")
+        $PSCmdlet.ThrowTerminatingError([System.Management.Automation.ErrorRecord]::new($exception, "FatalError", [Management.Automation.ErrorCategory]::InvalidOperation, $null))
+    }
+
+    $signature = Get-AuthenticodeSignature $asset
+
+    if ($signature.Status -ne "Valid")
+    {
+        return $false
+    }
+
+    if ($signature.SignerCertificate.Subject -ne 'CN="Datadog, Inc", O="Datadog, Inc", L=New York, S=New York, C=US')
+    {
+        return $false
+    }
+
+    if ($signature.SignerCertificate.Issuer -ne 'CN=DigiCert Trusted G4 Code Signing RSA4096 SHA384 2021 CA1, O="DigiCert, Inc.", C=US')
```

Comment:
Suggest being a little less restrictive on the issuer; concern is that when we get a new signing cert, the issuer could change in subtle ways (I'm thinking the year as an obvious one), which would cause the script to just start failing.

Reply:
The concern is valid - I think we can mitigate it through manual and automated tests that will identify a mismatch. Since we're using this for validation that our module hasn't been tampered with, the more restrictive the better.

We should ensure we have automated tests for this AND this install method should always be included during manual QA. I think those 2 things are enough to enable us to keep this restrictive.

@ISauve is this something that we can add into automated tests?

@clarkb7 how much of a pain would including this into every release QA cycle be?

@derekwbrown if you still feel we need to loosen the restriction, what would you suggest?

Reply:
Including in every cycle doesn't help.  Someone installs the module today.  Tomorrow we get a new cert that's slightly different.  then they run the command.  breaks for no reason.

So, to answer your questoin, I can think of a bunch of ways.
1) Least restrictive... as long as the signature is valid and has a valid chain, go for it.
2) More restrictive than (1), but more work.  Parse the CN of the cert, and make sure it includes (case insensitive) datadog.  Parse the CN of the issuer, make sure it  includes DigiCert.  Make sure the Object (O) includes digicert.

Even restricting to `DigiCert` is at least potentially flakey, if we ever change what CA we buy our certs from.

The OS is supposed to manage the trusted roots; tbh I think a valid signature is sufficient.